File: src/leetcode/15threeSum.cpp
// 这道题最重要的是学会去重
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

/* ------------------------------------------------------- 相向双指针 ------------------------------------------------------- */
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        int n = nums.size();
        vector<vector<int>> result;
        sort(nums.begin(), nums.end());
        for (int i = 0; i < n; ++i) {
            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            }
            int target = -nums[i];
            int right = n - 1;
            for (int left = i + 1; left < n; ++left) {
                if (left > i + 1 && nums[left] == nums[left - 1]) {
                    continue;
                }
                while (left < right && nums[left] + nums[right] > target) {
                    right--;
                }
                if (left == right) {
                    break;
                }
                if (nums[left] + nums[right] == target) {
                    result.push_back({nums[i], nums[left], nums[right]});
                }
            }
        }
        return result;
    }
};
int main(int argc, char const *argv[])
{
    vector<int> nums = {1, 0, -1, 0, -2, 2};
    Solution su;
    vector<vector<int>> result = su.threeSum(nums);
    for (int i = 0; i < result.size(); i++)
    {
        for (int j = 0; j < result[i].size(); j++)
        {
           cout << result[i][j] << " ";
        }
    }
    return 0;
}

File: scripts/leetcode/15threeSum.py
from typing import List
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        result = []
        nums.sort()
        for i in range(n):
            if nums[i] > 0:
                return result
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            target = -nums[i]
            right = n - 1
            for left in range(i + 1, n):
                if left > i + 1 and nums[left] == nums[left - 1]:
                    continue
                while left < right and nums[left] + nums[right] > target:
                    right -= 1
                if left == right:
                    break
                if nums[left] + nums[right] == target:
                    result.append([nums[i], nums[left], nums[right]])
        return result



s = Solution()
nums = [-1,0,1,2,-1,-4]
print(s.threeSum(nums))